### PR TITLE
[#11843] Add scheduled post processing step for student activity logs

### LINKED ISSUE #11843: Add scheduled post-processing step for student activity logs
**Problem statement**

For the student activity logs feature, each usage makes a call to Cloud Logging, which has QPS restriction as well as not-low latency. (a search space of 30 days can take 3-4 mins)

Such implementation is not suitable for a user-facing production feature:
- 3-4 mins latency without any kind of indicator is outright bad UX.
- The QPS restriction is going to haunt us in the future, if not right after the feature is released.
- In addition, Cloud Logging has a (default) bucket lifetime of 30 days.

**Solution**

The general idea is that we need to move the log entities to DB, while at the same time not directly writing to DB as:
- The transaction volume could be high, or at least higher than needed.
- We may not need to record every single activity, i.e. if there is someone who keeps refreshing the submission page, it is a waste to record (and also to display) every single so-called "page visit".

What we need to do is to regularly (e.g. one hour or 30-minute interval) process the logs and inserting the processed data to the DB as needed, similar to how the usage statistics are recorded. Afterwards, users can simply fetch data from the DB, which will be much faster or at least having trackable progress, no QPS restriction, and no expiry date.

### PR DESCRIPTION
**Pull request by daongochieu2810**
Fixes #11843 

**Outline of Solution**

1. Create db, logic and action class for new FeedbackSessionLogEntry entity for the datastore
2. Update `GetFeedbackSessionLogs` to get logs from datastore and remove 30-days restriction
3. Update `CreateFeedbackSessionLogs` to prevent logs created too close together (min 2s apart)
4. Create new cron job that runs every hour to get logs from gcloud logging service (filtered with the lastest timestamp in the logs from the datastore)

### CODE REVIEW

**Comment by samuelfangjw** [comment]
Apologies, I will try to review this as soon as possible but it seems like a decent sized PR and I might not have spare capacity to review this for awhile. @fsgmhoward, @zhaojj2209, requesting your help to review this if you have time.

File: src/e2e/java/teammates/e2e/cases/InstructorAuditLogsPageE2ETest.java [hunk 1]
```diff
@@ -80,6 +80,7 @@ public void testAll() {
 
         studentSubmissionPage.fillTextResponse(1, receiver.getName(), response);
         studentSubmissionPage.clickSubmitQuestionButton(1);
+        syncFeedbackSessionLogEntries();
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/InstructorAuditLogsPageE2ETest.java:83]: I really don't think this is needed. When run in production server, this will cause duplicated logs.

File: src/main/java/teammates/logic/api/Logic.java [hunk 2]
```diff
@@ -116,6 +119,16 @@ public List<String> updateReadNotifications(String googleId, String notification
         return accountsLogic.updateReadNotifications(googleId, notificationId, endTime);
     }
 
+    public List<FeedbackSessionLogEntryAttributes> getFeedbackSessionLogs(String courseId, String email,
+                                                                          long startTime, long endTime, String fsName) {
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/api/Logic.java:123]: Try not to apply an indentation where the parameters are squished all the way to the right like this. Applies everywhere in this PR.

File: src/main/java/teammates/logic/core/FeedbackSessionLogEntriesLogic.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    /**
+     * Gets the feedback session logs as filtered by the given parameters.
+     */
+    public List<FeedbackSessionLogEntryAttributes> getFeedbackSessionLogs(String courseId, String email,
+                                                                long startTime, long endTime, String fsName) {
+        return fslEntriesDb.getFeedbackSessionLogs(courseId, email, startTime, endTime, fsName);
+    }
+
+    /**
+     * Creates feed back session logs.
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/FeedbackSessionLogEntriesLogic.java:36]: feedback is one word on its own

File: src/main/java/teammates/logic/external/LocalLoggingService.java [hunk 4]
```diff
@@ -203,14 +204,25 @@ private boolean isRequestFilterSatisfied(LogDetails details, String actionClassF
...
         FEEDBACK_SESSION_LOG_ENTRIES.computeIfAbsent(courseId, k -> new ArrayList<>()).add(logEntry);
     }
 
     @Override
-    public List<FeedbackSessionLogEntry> getFeedbackSessionLogs(String courseId, String email,
-            long startTime, long endTime, String fsName) {
+    public List<FeedbackSessionLogEntryAttributes> getFeedbackSessionLogs(String courseId, String email,
+                                                                          long startTime, long endTime, String fsName) {
+        if (courseId == null) {
+            ArrayList<FeedbackSessionLogEntryAttributes> logEntries = new ArrayList<>();
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/external/LocalLoggingService.java:216]:
```suggestion
            List<FeedbackSessionLogEntryAttributes> logEntries = new ArrayList<>();
```

File: src/main/java/teammates/ui/webapi/CreateFeedbackSessionLogAction.java [hunk 5]
```diff
@@ -24,15 +24,16 @@ void checkSpecificAccessControl() {
...
         String fslType = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_LOG_TYPE);
         FeedbackSessionLogType convertedFslType = FeedbackSessionLogType.valueOfLabel(fslType);
         if (convertedFslType == null) {
             throw new InvalidHttpParameterException("Invalid log type");
         }
 
         String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
-        String fsName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
         String studentEmail = getNonNullRequestParamValue(Const.ParamsNames.STUDENT_EMAIL);
+        String fsName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/CreateFeedbackSessionLogAction.java:36]: All unrelated changes?

File: src/main/java/teammates/ui/webapi/Action.java [hunk 6]
```diff
@@ -86,6 +86,10 @@ public void setLogsProcessor(LogsProcessor logsProcessor) {
         this.logsProcessor = logsProcessor;
     }
 
+    public void setLogic(Logic logic) {
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/Action.java:89]: Why is this needed?

File: src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            String logType = logEntry.getFeedbackSessionLogType();
+            Long logTimestamp = logEntry.getTimestamp();
+            Map<String, Long> studentLog = studentLatestLogs.getOrDefault(studentEmail, new HashMap<>());
+            boolean isValid = true;
+
+            if (studentLog.containsKey(logType)) {
+                Long lastStudentLogTimestamp = studentLog.get(logType);
+
+                if (Math.abs(lastStudentLogTimestamp - logTimestamp) < 2 * 1000) {
+                    isValid = false;
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java:41]: Logs with close timestamps should not be duplicated, but we still want to know the number of occurrences (e.g. within 2 seconds, student A tries to submit 4 times).

File: src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        Map<String, Map<String, Long>> studentLatestLogs = new HashMap<>();
+        List<FeedbackSessionLogEntryAttributes> validLogEntries = new ArrayList<>();
+        List<FeedbackSessionLogEntryAttributes> allLogEntries =
+                logsProcessor.getFeedbackSessionLogs(null, null,
+                        currentTime.minus(15, ChronoUnit.MINUTES).toEpochMilli(), currentTime.toEpochMilli(), null);
+
+        for (FeedbackSessionLogEntryAttributes logEntry : allLogEntries) {
+            String studentEmail = logEntry.getStudentEmail();
+            String logType = logEntry.getFeedbackSessionLogType();
+            Long logTimestamp = logEntry.getTimestamp();
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java:33]:
```suggestion
            long logTimestamp = logEntry.getTimestamp();
```
I don't think timestamp is ever null?

File: src/test/java/teammates/logic/api/LogicExtension.java [hunk 9]
```diff
@@ -35,4 +36,10 @@ public List<FeedbackResponseAttributes> getFeedbackResponsesForQuestion(String f
         return feedbackResponsesLogic.getFeedbackResponsesForQuestion(feedbackQuestionId);
     }
 
+    @Override
+    public List<FeedbackSessionLogEntryAttributes> getFeedbackSessionLogs(String courseId, String email,
+                                                                             long startTime, long endTime, String fsName) {
+        return feedbackSessionLogEntriesLogic.getAllFeedbackSessionLogs();
```
Comment (wkurniawan07) [inline src/test/java/teammates/logic/api/LogicExtension.java:42]: Why do you need to override the method like this?

File: src/test/java/teammates/test/AbstractBackDoor.java [hunk 10]
```diff
@@ -869,6 +869,13 @@ public DeadlineExtensionAttributes getDeadlineExtension(
                 .build();
     }
 
+    /**
+     * Sync feedback session logs from GCloud Logging Service.
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/AbstractBackDoor.java:873]: "GCloud Logging Service" is an implementation detail that does not matter for BackDoor. It should not be mentioned here.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
There are unfortunately still many questionable parts.

File: src/e2e/java/teammates/e2e/cases/InstructorAuditLogsPageE2ETest.java (hunk 1 above)
Comment (daongochieu2810) [reply src/e2e/java/teammates/e2e/cases/InstructorAuditLogsPageE2ETest.java:83]: Im not sure what is better here? we need to check that the logs are correctly recorded, and in this case manually triggering the cron job seems needed to me

File: src/main/java/teammates/ui/webapi/CreateFeedbackSessionLogAction.java (hunk 5 above)
Comment (daongochieu2810) [reply src/main/java/teammates/ui/webapi/CreateFeedbackSessionLogAction.java:36]: We need `fsName` for `FeedbackSessionLogEntryAttributes` in `createFeedbackSessionLog` of `LocalLoggingService`

File: src/main/java/teammates/logic/core/FeedbackSessionLogEntriesLogic.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    public List<FeedbackSessionLogEntryAttributes> createFeedbackSessionLogs(
+            List<FeedbackSessionLogEntryAttributes> entries) throws InvalidParametersException {
+        return fslEntriesDb.createFeedbackSessionLogs(entries);
+    }
+
+    /**
+     * Gets all feedback session logs.
+     */
+    public List<FeedbackSessionLogEntryAttributes> getAllFeedbackSessionLogs() {
+        return fslEntriesDb.getAllFeedbackSessionLogs();
```
[the remark this answers is not in the document]
Comment (daongochieu2810) [reply src/main/java/teammates/logic/core/FeedbackSessionLogEntriesLogic.java:47]: I only use this for unit testing

File: src/main/java/teammates/ui/webapi/Action.java (hunk 6 above)
Comment (daongochieu2810) [reply src/main/java/teammates/ui/webapi/Action.java:89]: I need to use `getFeedbackSessionLogs` from LogicExtension (which use `getAllFeedbackSessionLogs`)

File: src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java (hunk 7 above)
Comment (wkurniawan07) [reply src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java:41]: You need to keep either the first or the last log and also store the info that there are 4 such attempts in the 2 seconds span.

Comment (daongochieu2810) [reply src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java:41]: > also store the info that there are 4 such attempts in the 2 seconds span.

How do we store this info?

Comment (wkurniawan07) [reply src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java:41]: I can give an answer, but I'll be doing your work if I do.

File: src/e2e/java/teammates/e2e/cases/InstructorAuditLogsPageE2ETest.java (hunk 1 above)
Comment (daongochieu2810) [reply src/e2e/java/teammates/e2e/cases/InstructorAuditLogsPageE2ETest.java:83]: @wkurniawan07 also I dont see why this would cause duplicated logs?

Comment (wkurniawan07) [reply src/e2e/java/teammates/e2e/cases/InstructorAuditLogsPageE2ETest.java:83]: Let's take a timespan from 10 - 11am. Logs collected by cron jobs are at 10.00, 10.15, 10.30, 10.45 (give/take 1-2 mins).
Run this E2E test at 10.20. Which logs is this test entering to the DB?

Comment (daongochieu2810) [reply src/e2e/java/teammates/e2e/cases/InstructorAuditLogsPageE2ETest.java:83]: ok I see, but how do we verify the presence of the created logs? Do we make a call to gcloud logging service directly?

File: src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java (hunk 7 above)
Comment (daongochieu2810) [reply src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java:41]: @wkurniawan07 do the 4 logs have to be of 1 type for them to be invalid? I was only considering close logs of the same type to be invalid, as @damithc instructed

Comment (damithc) [reply src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java:41]: Yes, my view is that rapid repeating of the same action is a 'possible spam' (best not to call them 'invalid'). If the user clicks submission links of two sessions rapidly, that is unlikely a spam. e.g.. the user could be opening links received in a 'links recovery' email.

File: src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+                    FeedbackSessionLogEntryAttributes startLog = logs.get(windowStartIndex);
+                    FeedbackSessionLogEntryAttributes currLog = logs.get(i);
+
+                    if (currLog.getTimestamp() - startLog.getTimestamp() <= MIN_WINDOW_PERIOD) {
+                        windowSize++;
+
+                        // If the window size exceeds the max value
+                        // we only take the first log of the window.
+                        if (windowSize == MAX_WINDOW_SIZE) {
+                            validLogEntries.add(startLog);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java:74]: This may prevent excessive logs, but the fact that there are 4 (magic number, by the way) accesses is still not recorded.

File: src/test/java/teammates/ui/webapi/GetFeedbackSessionLogsActionTest.java [hunk 13]
```diff
@@ -189,4 +197,21 @@ protected void testAccessControl() {
...
     }
 
+    private void validateFeedbackSessionLogOutput(
+            List<FeedbackSessionLogEntryData> outputEntries,
+            List<FeedbackSessionLogEntryAttributes> inputEntries
+    ) {
+        for (FeedbackSessionLogEntryData entryData : outputEntries) {
+            String logType = entryData.getFeedbackSessionLogType().getLabel();
+
+            assertTrue(
```
Comment (wkurniawan07) [inline src/test/java/teammates/ui/webapi/GetFeedbackSessionLogsActionTest.java:207]: If you're trying to assert that the lists are equal, this is not right. This here is checking whether `inputEntries` is a superset of `outputEntries`.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/e2e/java/teammates/e2e/cases/InstructorStudentActivityLogsPageE2ETest.java [hunk 14]
```diff
@@ -90,6 +93,7 @@ public void testAll() {
         studentActivityLogsPage.setActivityType("session access and submission");
         studentActivityLogsPage.waitForPageToLoad();
         studentActivityLogsPage.startSearching();
+        studentActivityLogsPage.addLogEntry(logEntry.getFeedbackSessionName());
```
[the remark this answers is not in the document]
Comment (daongochieu2810) [reply src/e2e/java/teammates/e2e/cases/InstructorStudentActivityLogsPageE2ETest.java:96]: Because the logs are not immediately available after the above actions, so I need to manually add the log

File: src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java (hunk 12 above)
Comment (daongochieu2810) [reply src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java:74]: should we add a new field `remarks` to the fb session log to hold this event?

File: src/e2e/java/teammates/e2e/cases/InstructorStudentActivityLogsPageE2ETest.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/e2e/java/teammates/e2e/cases/InstructorStudentActivityLogsPageE2ETest.java:96]: Then you are not testing anything of value in here. Think about it: the assertion is testing this line, not the actual logging function.

[the remark this answers is not in the document]
Comment (daongochieu2810) [reply src/e2e/java/teammates/e2e/cases/InstructorStudentActivityLogsPageE2ETest.java:96]: @wkurniawan07 so should we remove this test?

**Comment by hhdqirui** [comment]
@daongochieu2810 reminder on this PR thanks!

**Comment by fsgmhoward** [comment]
@daongochieu2810 Any updates for this PR?

**Comment by ziqing26** [comment]
@daongochieu2810 Another reminder for you on this PR, thank you!

**Comment by hhdqirui** [comment]
@daongochieu2810 kind reminder on this PR

**Comment by wkurniawan07** [comment]
@daongochieu2810 let us continue on the work here

**Comment by daongochieu2810** [comment]
@wkurniawan07 I have resolved your previous comments. Let me know if the current PR is ok and I can write more test cases for the log update cron job

File: src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+                        // If the window size exceeds the max value
+                        // we only take the first log of the window.
+                        if (windowSize == MAX_WINDOW_SIZE) {
+                            startLog.setRemarks("This window has " + windowSize + " log(s)");
+                            validLogEntries.add(startLog);
+                            windowStartIndex = i + 1;
+                        }
+                    } else {
+                        windowSize = 1;
+                        validLogEntries.add(logs.get(windowStartIndex));
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java:79]: I believe this can be just:
```suggestion
                        validLogEntries.add(startLog);
```

File: src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+
+                for (int i = 0; i < logs.size(); i++) {
+                    FeedbackSessionLogEntryAttributes startLog = logs.get(windowStartIndex);
+                    FeedbackSessionLogEntryAttributes currLog = logs.get(i);
+
+                    if (currLog.getTimestamp() - startLog.getTimestamp() <= MIN_WINDOW_PERIOD) {
+                        windowSize++;
+                        // If the window size exceeds the max value
+                        // we only take the first log of the window.
+                        if (windowSize == MAX_WINDOW_SIZE) {
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java:72]: Why do we need max window size in the first place? Recorded spammy/invalid actions shouldn't be limited. If someone can somehow access 1000 times in 2s, it should show all 1000.

File: src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+                            startLog.setRemarks("This window has " + windowSize + " log(s)");
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java:73]: This `remarks` field has some weird purpose.
- Why must it be a proper sentence? What is the downstream usage of this field?
  - In fact, why is it called `remarks`?
- Why not just record the number as is?

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Still some questionable parts.

I think, when you start writing the tests for the new action, you will be able to follow the logic clearer.

File: src/e2e/java/teammates/e2e/cases/InstructorStudentActivityLogsPageE2ETest.java [hunk 18]
```diff
@@ -83,6 +84,8 @@ public void testAll() {
 
         studentSubmissionPage.fillTextResponse(1, receiver.getName(), response);
         studentSubmissionPage.clickSubmitQuestionButton(1);
+        // Manually trigger the log update cron job
+        BACKDOOR.executeGetRequest(Const.CronJobURIs.AUTOMATED_FEEDBACK_SESSION_LOGS_UPDATE, new HashMap<>());
```
[the remark this answers is not in the document]
Comment (daongochieu2810) [reply src/e2e/java/teammates/e2e/cases/InstructorStudentActivityLogsPageE2ETest.java:88]: the logs are not available immediately in the db, so we need to trigger the cron job manually to have it ready before the next test

File: src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java (hunk 17 above)
Comment (daongochieu2810) [reply src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java:73]: `remarks` is to record abnormalities, and in this case, they are spams. We could just record the number, but I thought that might be confusing

File: src/e2e/java/teammates/e2e/cases/InstructorStudentActivityLogsPageE2ETest.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/e2e/java/teammates/e2e/cases/InstructorStudentActivityLogsPageE2ETest.java:88]: But you are not testing that in the first place. The logs are entered via data JSON file.

File: src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java (hunk 17 above)
Comment (wkurniawan07) [reply src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java:73]: `remarks` is far more confusing. No part of the field says it's for abnormality, and it already assumes too much of downstream use case.
Whereas if you record the number, there can be many usages, because without context it is just a number.

And of course, the field name should not be `remarks`. If it represents number of accesses, the field name should say it clearly.

**Comment by daongochieu2810** [comment]
@wkurniawan07 I have added the unit test for FeedbackSessionLogUpdateAction. This should ensure the correctness of the cron job logic

File: src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+                currentTime.minus(1, ChronoUnit.SECONDS).toEpochMilli(),
+                currentTime.toEpochMilli(), null);
+
+        assertEquals(1, logs.size());
+
+        FeedbackSessionLogEntryAttributes log = logs.get(0);
+
+        assertEquals("student1InCourse1@gmail.tmt", log.getStudentEmail());
+        assertEquals("First feedback session", log.getFeedbackSessionName());
+        assertEquals(FeedbackSessionLogType.ACCESS.getLabel(), log.getFeedbackSessionLogType());
```
Comment (wkurniawan07) [inline src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java:65]: Should check (1) the window size of the log attributes and (2) the timestamp (check that it's the earlier log that gets added).

File: src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+                FeedbackSessionLogEntryAttributes startLog = logs.get(0);
+
+                for (FeedbackSessionLogEntryAttributes currLog : logs) {
+                    if (currLog.getTimestamp() - startLog.getTimestamp() < MIN_WINDOW_PERIOD) {
+                        windowSize++;
+                    } else {
+                        // Take the first log of the window.
+                        if (windowSize > MAX_WINDOW_SIZE) {
+                            startLog.setWindowSize(windowSize);
+                        }
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/FeedbackSessionLogsUpdateAction.java:66]: - I don't believe you need to define` MAX_WINDOW_SIZE` explicitly. The condition is simply "if there are multiple logs", not "if there is > x amount of logs, where x = 1".
- In fact, this check should not even be needed, similarly for down there. `startLog.setWindowSize(1)` is itself accurate.

File: src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+
+        // wait 5s to isolate the previous log window
+        ThreadHelper.waitFor(5 * 1000);
```
Comment (wkurniawan07) [inline src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java:68]: Every place where `5 * 1000` and `2 * 1000` are defined in this class are magic numbers. They are based on a certain field in the class being tested.

File: src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+    protected void testAccessControl() {
+        // verifyOnlyAdminCanAccess();
+    }
+
+    @Override
+    @Test
+    public void testExecute() throws Exception {
+        FeedbackSessionLogsUpdateAction action = getAction();
+
+        ______TS("creating 2 logs in the same window should return only 1 log");
```
Comment (wkurniawan07) [inline src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java:43]: The fact that "2" can be mentioned here and many other places down means that `MAX_WINDOW_SIZE` is really redundant.

File: src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+import teammates.common.util.Const;
+import teammates.test.ThreadHelper;
+
+/**
+ * SUT: {@link FeedbackSessionLogsUpdateAction}.
+ */
+public class FeedbackSessionLogsUpdateActionTest
+        extends BaseActionTest<FeedbackSessionLogsUpdateAction> {
+
+    private static final int MIN_WINDOW_PERIOD = 2 * 1000;
```
Comment (wkurniawan07) [inline src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java:20]: This should be read from the main class. Copying here risks de-synchronization if we are to modify this value.

File: src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+    private static final int ISOLATION_PERIOD = 5 * 1000;
```
Comment (wkurniawan07) [inline src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java:21]: This is a function of the `MIN_WINDOW_PERIOD`, so it shouldn't be hardcoded.

File: src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    @Override
+    @Test
+    public void testExecute() throws Exception {
+        FeedbackSessionLogsUpdateAction action = getAction();
+
+        ______TS("creating 2 logs in the same window should return only 1 log");
+
+        // create a window exceeding max size
```
Comment (wkurniawan07) [inline src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java:46]: There is no such thing as "max size" anymore. This is essentially creating multiple logs in one window.

File: src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+        assertEquals(2, allLogEntries.size());
+
+        FeedbackSessionLogEntryAttributes firstLog = logs.get(0);
+        FeedbackSessionLogEntryAttributes secondLog = allLogEntries.get(1);
+        long windowSize = secondLog.getTimestamp() - firstLog.getTimestamp();
+
+        // check that the first log is taken
+        // and the window size is smaller than the min value
+        assertTrue(windowSize > 0);
+        assertTrue(windowSize < MIN_WINDOW_PERIOD);
```
Comment (wkurniawan07) [inline src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java:75]: I'm not sure why testing this `windowSize` is needed. Checking that the first log is taken is already guaranteed by the first `assertEquals` after this, and checking the window size is nothing of value here, unless we are testing multiple windows.

File: src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java [hunk 27]
```diff
@@ -0,0 +1,145 @@
...
+
+        assertEquals(firstLog.getTimestamp(), allLogEntries.get(0).getTimestamp());
+        assertEquals("student1InCourse1@gmail.tmt", firstLog.getStudentEmail());
+        assertEquals("First feedback session", firstLog.getFeedbackSessionName());
+        assertEquals(FeedbackSessionLogType.ACCESS.getLabel(), firstLog.getFeedbackSessionLogType());
+
+        // wait 5s to isolate the previous log window
```
Comment (wkurniawan07) [inline src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java:82]: "Wait 5s": a magic number is being mentioned here

File: src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java [hunk 28]
```diff
@@ -0,0 +1,145 @@
...
+        FeedbackSessionLogEntryAttributes accessLog1 = logs.get(0);
+        FeedbackSessionLogEntryAttributes accessLog2 = logs.get(1);
+
+        assertEquals("student1InCourse1@gmail.tmt", accessLog1.getStudentEmail());
+        assertEquals("First feedback session", accessLog1.getFeedbackSessionName());
+        assertEquals(FeedbackSessionLogType.ACCESS.getLabel(), accessLog1.getFeedbackSessionLogType());
+
+        assertEquals("student1InCourse1@gmail.tmt", accessLog2.getStudentEmail());
+        assertEquals("First feedback session", accessLog2.getFeedbackSessionName());
+        assertEquals(FeedbackSessionLogType.ACCESS.getLabel(), accessLog2.getFeedbackSessionLogType());
```
Comment (wkurniawan07) [inline src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java:142]: The assertion is identical for `accessLog1` and `accessLog2`. How are we sure that they're referring to the correct ones?

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java (hunk 26 above)
Comment (daongochieu2810) [reply src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java:75]: this is to check that the 2nd log was discarded correctly i.e the gap was larger than the min period

Comment (wkurniawan07) [reply src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java:75]: If this is checking the 2nd log was discarded, then what does this line do:
```java
assertEquals(1, logs.size());
```

Comment (daongochieu2810) [reply src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java:75]: the above line check that the 2nd log was discarded, line 74 and 75 verify that it was discarded **correctly**

Comment (wkurniawan07) [reply src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java:75]: I still don't get it. What cases can fail here that all other parts of this test will not fail?

Comment (daongochieu2810) [reply src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java:75]: lets say for some reason, these 2 logs are 3 seconds apart, but only 1 log was returned. In this case, if we dont include line 74 and 75, the test will pass which is not what we want

Comment (wkurniawan07) [reply src/test/java/teammates/ui/webapi/FeedbackSessionLogsUpdateActionTest.java:76]: > these 2 logs are 3 seconds apart, but only 1 log was returned

If the later test "creating 2 logs in different windows should return both logs" does its job, this is an unreachable scenario.

**Comment by daongochieu2810** [comment]
@wkurniawan07 I have resolved all of the issues, thanks!

**Comment by hhdqirui** [comment]
Hi @daongochieu2810 please also fix the component tests, thanks!

**Comment by fsgmhoward** [comment]
I saw no updates for this PR and I assume it is ready for review? @hhdqirui @samuelfangjw Please take a look at it. Thanks.

**Comment by wkurniawan07** [comment]
Will be closing this for now, we will revisit the feature in greater detail in a few weeks time
